feat: Job String Processor

merlin-northern opened the pull request:
Changelog: Title
Ticket: MEN-5819
Signed-off-by: Peter Grzybowski <peter@northern.tech>

merlin-northern commented:
@mender-test-bot start pipeline please

merlin-northern commented:
@mender-test-bot start pipeline please

merlin-northern commented:
@mender-test-bot start pipeline please

merlin-northern commented:
@tranchitella could I kindly ask you to take a look, since you already reviewed the first try (the quck fix from the past weekend in ent)

tranchitella commented on app/worker/process.go:
```diff
@@ -142,25 +144,23 @@ func processTask(task model.Task, job *model.Job,
 				"Error: Task definition incompatible " +
 					"with specified type (http)")
 		}
-		l.Infof("processTask: calling http task: %s %s", httpTask.Method,
-			processJobString(httpTask.URI, workflow, job))
```
Why did you remove this log?

merlin-northern replied:
1. I moved it below
2. it was doubling the very expensive operation of processing the job string, just for printing.

tranchitella commented on app/worker/http.go:
```diff
@@ -39,33 +40,42 @@ var makeHTTPRequest = func(req *http.Request, timeout time.Duration) (*http.Resp
+) (*model.TaskResult, error) {
+	options := &processor.Options{Encoding: processor.URL}
+	uri := ps.ProcessJobString(httpTask.URI, options)
+	l.Infof("processHTTPTask: starting with: method=%s uri=%s (options=%+v)",
+		httpTask.Method,
+		uri,
+		options,
+	)
```
I don't think logging option is valuable, as it will always be Encoding: URL.
By the way, I don't get why we need this log here.

merlin-northern replied:
I can remove it, it was moved from one level up (see the previous comment). do you want me to remove it?

tranchitella commented on app/worker/http.go:
```diff
@@ -39,33 +40,42 @@ var makeHTTPRequest = func(req *http.Request, timeout time.Duration) (*http.Resp
-			value = processJobString(value, workflow, job)
-			value = maybeExecuteGoTemplate(value, job.InputParameters.Map())
+			key = ps.ProcessJobString(key, nil)
+			key = maybeExecuteGoTemplate(key, ps.GetJobInputParameters().Map())
```
It looks weird obtaining the job's input parameters from a `GetJobInputParameters` method on the JobStringProcessor instance. It looks like you did it by convenience because processHTTPTask doesn't receive anymore the job, but honestly I don't like it too much.

merlin-northern replied:
I know, but there is no good way of going it. would you prefer `ps.GetJob()` method instead?

merlin-northern replied:
@alfrunes there is a bit of refactoring going on here (I could not stop) in order not to redesign the whole thing, I added `GetJobInputParameters` in the string processor class. we (Fabio and me) both think it is not very elegant, and a bit out of place. what do you think? (Keep in mind that we do not want to go over board here :>)

alfrunes replied:
Since `maybeExecuteGoTemplate` is called wherever `ps.ProcessJobString` is called; wouldn't it be easier to just move it inside the `processor` package?

merlin-northern replied:
brilliant. so I created a method `MaybeExecuteGoTemplate` but there were complications, in the end, this is the final version. makes sense? beware: I can refactor deeper.

tranchitella commented on app/worker/process.go:
```diff
@@ -142,25 +144,23 @@ func processTask(task model.Task, job *model.Job,
 				"Error: Task definition incompatible " +
 					"with specified type (http)")
 		}
-		l.Infof("processTask: calling http task: %s %s", httpTask.Method,
-			processJobString(httpTask.URI, workflow, job))
```
why did you remove this log and moved it in HTTPTask?

merlin-northern replied:
it was super expensive to process the job string twice, here for only logging and then again right after.

tranchitella commented on app/processor/string.go:
```diff
@@ -0,0 +1,138 @@
+	}
+}
+
+func (j *JobStringProcessor) ProcessJobString(data string, options *Options) string {
```
If you use a variadic function parameter, you can save the `, nil` where you don't need otpions:

```suggestion
func (j *JobStringProcessor) ProcessJobString(data string, opts ...*Options) string {
```

At this point, you can loop over options (not tested):
```
    options := Options{Encoding: Plain}
    for _, o := range opts {
        options.Encoding = o.Encoding
    }
```

If you want to spend a bit more time, you can use the functional options pattern:
https://www.sohamkamani.com/golang/options-pattern/

merlin-northern replied:
I was following your previous suggestions :) I like the `...`. the functional options: nice! I have never used it. I think `...` is nice here. is it ok now?

merlin-northern replied:
will you believe it? `app/processor/string.go:67:1: cyclomatic complexity 21 of func `(*JobStringProcessor).ProcessJobString` is high (> 20) (gocyclo)` there is always something.

merlin-northern commented:
@mender-test-bot start pipeline please sugar

alfrunes commented:
Is this a `feat`ure? I would argue that this is a `refac`tor of existing functionality.

alfrunes commented on app/processor/job.go:
```diff
@@ -0,0 +1,85 @@
+	job *model.Job
+}
+
+type JsonOptions struct {
+}
```
I don't understand the purpose of this.

merlin-northern replied:
this is on purpose: to keep the interface consistent. in case there is a new round of refactoring (i.e.: a general processor interface) can we keep it?

alfrunes commented on app/processor/job.go:
```diff
@@ -0,0 +1,85 @@
+func (j JobProcessor) ProcessJSON(
+	data interface{},
+	ps *JobStringProcessor,
+	options ...*JsonOptions,
```
This variadic argument is unused.

merlin-northern replied:
this is on purpose: to keep the interface consistent. in case there is a new round of refactoring (i.e.: a general processor interface) can we keep it?

merlin-northern commented:
> Is this a `feat`ure? I would argue that this is a `refac`tor of existing functionality.

something in between feat, refact, fix, I will change to fix, since it fixes a bug. makes sense?

alfrunes commented on app/worker/http.go:
```diff
@@ -39,34 +40,41 @@ var makeHTTPRequest = func(req *http.Request, timeout time.Duration) (*http.Resp
+	l *log.Logger,
+) (*model.TaskResult, error) {
+	options := &processor.Options{Encoding: processor.URL}
+	uri := ps.ProcessJobString(httpTask.URI, options)
```
This is going to break certain URLs. For example if the url contains a `host:port` address. To be honest, I think the query escape should be done at the templating level and thus done as part of the workflow definition.

merlin-northern replied:
wait how can it be? it only URL-encodes the value:

[app/processor/string.go:97](https://github.com/mendersoftware/workflows/pull/234/files#diff-a395ff9022c3014278f32fad3c86b421391bc1558be6cf5ff6539e33cbbcbd89R97)
```go
						value = url.QueryEscape(value)
```

alfrunes replied:
This transforms all of the following characters: `'$', '&', '+', ',', '/', ':', ';', '=', '?', '@'` many of them are valid in different parts of the schema. For example
```go
url.QueryEscape("https://hosted.mender.io") // -> https%3A%2F%2Fhosted.mender.io
```

merlin-northern replied:
but the URL is never a workflow input variable, we take it only from env and with default values (`"uri": "http://${env.USERADM_ADDR|mender-useradm:8080}/api/internal/v1/useradm/tokens?tenant_id=${workflow.input.tenant_id}"`). this assumes that we encode the workflow input variables present in the URL (`${workflow.input.tenant_id}`).

merlin-northern replied:
ok, how about now? we support cool `encoding=plain|url;identifier` syntax that overrides everything you may pass. if you do not pass it, by default we URL encode all input variables that exist in the URL.

merlin-northern replied:
@tranchitella I hope you are happy that you used Alf as a mole to have this feature :>

merlin-northern commented:
@mender-test-bot start pipeline please sugar pretty please

alfrunes commented on app/processor/string.go:
```diff
@@ -0,0 +1,198 @@
+			len(match) > len(workflowEnvVariable) {
+			// Replace ${env.KEY} with the KEY environment variable
+			envName := match[len(workflowEnvVariable):]
+			envValue := os.Getenv(envName)
+			if envValue == "" {
+				envValue = defaultValue
+			}
+			data = strings.ReplaceAll(data, submatch[0], envValue)
```
```suggestion
			if envValue := os.Getenv(envName); envValue != "" {
				value = envValue
			}
```

alfrunes commented on app/processor/string.go:
```diff
@@ -0,0 +1,198 @@
+						varValue = ""
+					}
+					varValueString, err := utils.ConvertAnythingToString(varValue)
+					if err == nil {
+						if varValueString == "" {
+							varValueString = defaultValue
+						}
+						data = strings.ReplaceAll(data, submatch[0], varValueString)
+					}
```
```suggestion
					if err == nil {
						if varValueString != "" {
							value = varValueString
						}
					} else {
						continue SubmatchLoop
					}
```

alfrunes commented on app/processor/string.go:
```diff
@@ -0,0 +1,198 @@
+	option := getOption(options...)
+
+	// search for ${...} expressions in the data string
+	for _, submatch := range matches {
```
```suggestion
SubmatchLoop:
	for _, submatch := range matches {
```

alfrunes commented on app/processor/string.go:
```diff
@@ -0,0 +1,198 @@
+					break
+				}
+			}
+		}
```
Finally, this applies the option to all the replacements above.
```suggestion
		}
		if option.Encoding == URL {
			value = url.QueryEscape(value)
		}
		data = strings.ReplaceAll(data, submatch[0], value)
```

merlin-northern commented:
@alfrunes I moved to this PR: https://github.com/mendersoftware/workflows/pull/235 I fixed the units, and added tests for a new `processOptionString`, I had to rewrite that, ti was not working.

merlin-northern commented:
this one went with https://github.com/mendersoftware/workflows/pull/235
